src/DependencyInjection/DI/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs

```diff
@@ -49,25 +56,71 @@ private class ILEmitResolverBuilderRuntimeContext
             _runtimeResolver = runtimeResolver;
             _serviceScopeFactory = serviceScopeFactory;
             _rootScope = rootScope;
+            _scopeResolverCache = new ConcurrentDictionary<ServiceCacheKey, GeneratedMethod>();
         }
 
         public Func<ServiceProviderEngineScope, object> Build(ServiceCallSite callSite)
         {
-            if (callSite.Cache.Location == CallSiteResultCacheLocation.Root)
-            {
-                // If root call site is singleton we can return Func calling
-                // _runtimeResolver.Resolve directly and avoid Expression generation
-                if (TryResolveSingletonValue(callSite, out var value))
-                {
-                    return scope => value;
-                }
+            return BuildType(callSite).Lambda;
+        }
 
-                return scope => _runtimeResolver.Resolve(callSite, scope);
+        private GeneratedMethod BuildType(ServiceCallSite callSite)
+        {
+            // Only scope methods are cached
```

Comment:
I'm probably missing something obvious, but why not cache GeneratedMethods for transient services?

Reply:
I think I remember now. It's because the methods that don't need to check the ResolvedServices dictionary are already small enough. Still, if there's no perf impact of splitting scoped service instantiation into individual methods, it might be worth trying for non-scoped services.

Reply:
My current thinking is that there is no perf impact because scope resolution methods themselves are comparably slow. There is also not much benefit in caching transient resolvers because most of them contain a single method call anyway.